shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/QueueDispatcher.kt

```diff
@@ -0,0 +1,154 @@
+package io.customer.shared.tracking.queue
+
+import io.customer.shared.database.TrackingTask
+import io.customer.shared.database.TrackingTaskQueryHelper
+import io.customer.shared.sdk.config.BackgroundQueueConfig
+import io.customer.shared.tracking.api.*
+import io.customer.shared.tracking.api.model.*
+import io.customer.shared.tracking.constant.Priority
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.util.DateTimeUtil
+import io.customer.shared.util.Logger
+import io.customer.shared.work.*
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.datetime.Instant
+import kotlin.time.DurationUnit
+
+/**
+ * Class responsible for running queue operations. This class only receives requests to validate the
+ * queue and is responsible for following:
+ *
+ * - whether to run the queue now or wait (except when specifically asked to run).
+ * - maintain lock to avoid sending same tasks multiple times.
+ * - validate and update task status related to queue operations.
+ * - schedule timer for next queue request.
+ */
+internal interface QueueDispatcher {
+    suspend fun checkForPendingTasks(source: QueueTriggerSource)
+    suspend fun sendAllPending()
+}
+
+internal class QueueDispatcherImpl(
+    private val logger: Logger,
+    private val dateTimeUtil: DateTimeUtil,
+    override val jobExecutor: JobExecutor,
+    private val backgroundQueueConfig: BackgroundQueueConfig,
+    private val trackingTaskQueryHelper: TrackingTaskQueryHelper,
+    private val queueRunner: QueueRunner,
+    private val queueTimer: QueueTimer,
+) : QueueDispatcher, JobDispatcher {
+    private val mutex = Mutex()
+
+    private suspend fun acquireLock() {
+        logger.debug("Acquiring lock for queue")
+        kotlin.runCatching {
+            mutex.lock(owner = this)
```

Comment:
This is the only lock we have for tasks queue, I'm trying to test if we can only rely on database lock and exclude this one too, but that will take some time and is something we can keep for later.